doc/developer/design/20230329_monotonic_optimization.md

```diff
@@ -0,0 +1,177 @@
+
+# Motivation
+[motivation]: #motivation
+
+<!--
+Why are we doing this? What problems does it solve? What use cases does it enable? What is the desired outcome?
+-->
+
+Materialize currently focuses on rendering a good plan for workloads that need incremental view maintenance (IVM).  This misses opportunities to optimize plans based on the time bounds of a dataflow, which are in the case for `SELECT` queries a single timestamp.  In this design, we present an updated approach for the optimizer to handle monotonicity information, and to tighten the contract between plans and how they're rendered.
+
+# Explanation
+[explanation]: #explanation
+
+<!--
+Explain the design as if it were part of Materialize and you were teaching the team about it.
+This can mean:
+
+- Introduce new named concepts.
+- Explain the feature using examples that demonstrate product-level changes.
+- Explain how it builds on the current architecture.
+- Explain how engineers and users should think about this change, and how it influences how everyone uses the product.
+- If needed, talk though errors, backwards-compatibility, or migration strategies.
+- Discuss how this affects maintainability, or whether it introduces concepts that might be hard to change in the future.
+-->
+
+To start, we introduce the following monotonicity variants:
+* *Non-monotonic*: The data can contain retractions over time.
+* *Physically monotonic*: The data never contains retractions.
+* *Logically monotonic*: The data can contain retractions, but once consolidated it is physically monotonic.
+
+At the moment, the optimizer thinks in terms of the first two variants, but rendering assumes logical monotonicity.
+
+We need to change the following parts of Materialize:
+
+1. The optimizer derives monotonicity information based on the inputs and MIR expressions.  It does not take time bounds into consideration.  Here, it needs to flag inputs as logically monotonic if the dataflow only consumes a single point in time. This is partially achieved by MaterializeInc/materialize#14607.
+2. The [monotonic flag calculation](../../../src/transform/src/monotonic.rs) needs to distinguish logical vs. physical monotonicity.  It seems we need to extend the optimizer to actively adjust plans based on monotonicity information and requirements.
+3. LIR operators need to define what kind of monotonicity they can handle and produce.
+4. LIR operators must not make assumptions about the monotonicity characteristics of the data but should expect their requirements to be fulfilled.  This means that operators requiring physical monotonicity should not by default insert a defensive consolidation.
```

Comment:
nit: I think the two parts of the first sentence

> LIR operators must not make assumptions about the monotonicity characteristics of the data 
and
> LIR operators [...] should expect their requirements to be fulfilled

somehow contradict each other.